src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,172 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::credentials::Result;
+use crate::credentials::internal::sts_exchange::ClientAuthentication;
+use crate::headers_util::build_bearer_headers;
+use crate::token::{CachedTokenProvider, Token, TokenProvider};
+use crate::token_cache::TokenCache;
+use gax::error::CredentialsError;
+use http::{Extensions, HeaderMap};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::time::{Duration, Instant};
+
+use super::Credentials;
+use super::dynamic::CredentialsProvider;
+use super::external_account_sources::url_sourced_account::UrlSourcedCredentials;
+use super::internal::sts_exchange::{ExchangeTokenRequest, STSHandler};
+
+#[async_trait::async_trait]
+pub(crate) trait SubjectTokenProvider: std::fmt::Debug + Send + Sync {
+    /// Generate subject token that will be used on STS exchange.
+    async fn subject_token(&self) -> Result<String>;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceFormat {
+    #[serde(rename = "type")]
+    pub format_type: String,
+    pub subject_token_field_name: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceHeaders {
+    #[serde(flatten)]
+    pub headers: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSource {
+    pub url: Option<String>,
+    pub headers: Option<CredentialSourceHeaders>,
+    pub format: Option<CredentialSourceFormat>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ExternalAccountConfig {
+    pub audience: String,
+    pub subject_token_type: String,
+    pub token_url: String,
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub credential_source: CredentialSource,
+}
+
+#[derive(Debug)]
+struct ExternalAccountTokenProvider {
+    subject_token_provider: Box<dyn SubjectTokenProvider>,
+    config: ExternalAccountConfig,
+}
+
+impl ExternalAccountTokenProvider {
+    pub fn new(config: ExternalAccountConfig) -> Result<ExternalAccountTokenProvider> {
+        let subject_token_provider = subject_token_provider_from_config(config.clone())?;
+        Ok(Self {
+            subject_token_provider,
+            config,
+        })
+    }
+}
+
+/// Detect which subject token provider implementation to use
+fn subject_token_provider_from_config(
+    config: ExternalAccountConfig,
+) -> Result<Box<dyn SubjectTokenProvider>> {
+    if let Some(url) = config.credential_source.url {
+        let creds = UrlSourcedCredentials {
+            url,
+            headers: config.credential_source.headers,
+            format: config.credential_source.format,
+        };
+        return Ok(Box::new(creds));
```

Comment:
The creds type detection logic can move to the external account builder and maybe we could get away with `dyn SubjectTokenProvider`?